### PR TITLE
Add async option

### LINKED ISSUE #8200: Re-implement async config option
### What problem does this feature solve?
Vue Test Utils has a synchronous API by walking all watchers and setting them to `sync`, but this implementation is buggy.

Reimplementing an async private config option would solve this problem for Vue Test Utils.

### What does the proposed API look like?
Add an `async` config option that would be used when watchers are created.

The same option that was removed in [this commit](https://github.com/vuejs/vue/commit/ad11aa122cd09a4510fd9d8829e25f4a90b178c2#diff-4d10b522e21ed424211799a4e11c6103L191).

I'm happy to make a PR to add this.

### PR DESCRIPTION
**Pull request by eddyerburgh**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

**Other information:**

Closes #8200

We can use this in Vue Test Utils to run tests synchronously without editing Vue internals

### CODE REVIEW

**Comment by posva** [comment]
Do you think it would make sense to use an ENV variable so it is completely stripped out in normal builds?

**Comment by trollepierre** [comment]
what's up with this amazing feature?

File: src/core/observer/dep.js [hunk 1]
```diff
@@ -36,6 +37,12 @@ export default class Dep {
   notify () {
     // stabilize the subscriber list first
     const subs = this.subs.slice()
+    if (process.env.NODE_ENV !== 'production' && !config.async) {
```
Comment (bost-h) [inline src/core/observer/dep.js]: Is it necessary to check for `NODE_ENV` ? I usually have different tests configurations, and one of them sets `NODE_ENV` to production to be as close as possible to the production behaviour. By reading this code, I guess my tests won't work.

**Comment by sobrinho** [comment]
Waiting for that too, Vuetify is near impossible to test on vue-test-utils.

I'm seeing errors related to _transitionClasses not defined everywhere.

**Comment by benirose** [comment]
Is this still going into 2.6?

**Comment by eddyerburgh** [comment]
It should be out in 2.5.18

**Comment by aldarund** [comment]
@eddyerburgh is there any ETA on 2.5.18 ?

**Comment by agualis** [comment]
> @eddyerburgh is there any ETA on 2.5.18 ?
🎉 🎉 
https://github.com/vuejs/vue/releases/tag/v2.5.18-beta.0

**Pull request merged**